text/0000-elementary-functions.md

```diff
@@ -0,0 +1,122 @@
+in general.
+
+In order to prevent forks because of different results on different
+implementations or architectures, we provide our own implementation of the
+necessary approximation functions.
+
+Assuming the same representation of non-integral values and standardized basic
+arithmetic operations, this leads to equivalent results using different
+implementation languages or CPU architectures.
+
+# Decision
+[Decision]: #decision
+
+The library `non-integer` exports the functions `exp'` which approximates the
+exponential function, `ln'` which approximates the natural logarithm and the
+infix operator `(***)` which approximates exponentiation.
+
+Current implementation: `dependencies/non-integer` of `fm-ledger-rules@727cca6`
+(cf. submodule pointer).
+
+All input parameters for the functions need to be instances of `RealFrac` in
+order to support addition, subtraction, multiplication and division. The
+approximation algorithm terminates if the basic operations terminate. The
+approximation is recursive, the recursion depth is bounded by both a fixed
+constant and the precision of succeeding incremental approximations.
+
+The infix `(***)` operator uses both `ln` and `exp` in the following way
+`a *** b = exp' (b * ln' a)`.
+
+# Technical explanation
+[technical-explanation]: #technical-explanation
+
+The core of the approximation of `exp` and `ln` are continued fractions. For
+each function, there exists a continued fraction representation (e.g.,
+[log](http://functions.wolfram.com/ElementaryFunctions/Log/10/),
+[exp](http://functions.wolfram.com/ElementaryFunctions/Exp/10/)). Continued
+fractions can easily be calculated recursive using via a recurrence
+relation which results in an intermediate approximation after each step.
+
+Continued fractions do have advantages over finite Taylor / MacLaurin polynomial
+approximations:
+
+* For `ln` the Taylor series converges very badly and has a very high
+  approximation error even within the interval [1,2].
+* Taylor approximation required calculation of large factorials
+
+The convergence for both approaches is best in small domains, therefore we use
+mathematical transformations to scale the input parameters accordingly.
+
+For `exp'` we use the fact that for an integer `n` we have `e^x = e^(x/n)^n` and
+that `e^(-x)=1/e^x`. We therefore ensure that the parameter `y = x/n`for the
+approximation is non-negative and scale it to [0,1] via `y = x/ceiling(x)`. The
+final integer exponentiation can be done using the fast integer exponentiation
+algorithm.
+
+For `ln'` the continued fraction approximation is best for `[1,c]` with some
+small `c`. We first calculate an integer `n` such that `e^n <= x < e^(n+1)`
+(i.e., `1 <= x/(e^n) < e`) using bisection. Using the identity `ln(x*y) =
+ln(x) + ln(y)` we get `ln(x) = ln(e^n * x / e^n) = ln(e^n) + ln(x/e^n)` where
+`ln(e^n) = n` and `x/e^n` is an element of `[1,e)`, for which the approximation
+converges well.
```

Comment:
all of this is not really helping to write the code. This part is more fitting for the Decision level. The technical explanation is what you have in that other file : [NonIntegral.hs](https://github.com/input-output-hk/fm-ledger-rules/blob/master/dependencies/non-integer/src/NonIntegral.hs).

I'll remind you that line: [It is reasonably clear how the feature would be implemented](https://github.com/input-output-hk/implementation-decisions/blob/master/XXXX-template.md#technical-explanation).

I know this is hard, but you write this not as an academic but as a software developer. We need to be very accessible here, to not leave any details unsaid : otherwise a developer might get this wrong and this could affect out crypto-currency ecosystem.

Reply:
@NicolasDP would you prefer pseudo-code here ?